packages/gen-ai-hub/src/utils/deployment-resolver.ts

```diff
@@ -0,0 +1,80 @@
+import {
+  DeploymentApi,
+  AiDeployment,
+  AiDeploymentStatus
+} from '@sap-ai-sdk/ai-core';
+
+/**
+ * A deployment resolver can be either a deployment ID or a function that returns a full deployment object.
+ */
+export type DeploymentResolver = DeploymentId | (() => Promise<AiDeployment>);
+/**
+ * A deployment ID is a string that uniquely identifies a deployment.
+ */
+export type DeploymentId = string;
+/**
+ * A foundation model is identifier by its name and potentially a version.
+ */
+export interface FoundationModel {
+  /**
+   * The name of the model.
+   */
+  name: string;
+  /**
+   * The version of the model.
+   */
+  version?: string;
+}
+
+/**
+ * Query the AI Core service for a deployment that matches the given criteria. If more than one deployment matches the criteria, the first one is returned.
+ * @param opts - The options for the deployment resolution.
+ * @param opts.scenarioId - The scenario ID of the deployment.
+ * @param opts.executableId - The executable of the deployment.
+ * @param opts.model - The name and potentially version of the model to look for.
+ * @returns An AiDeployment, if a deployment was found, fails otherwise.
+ */
+export async function resolveDeployment(opts: {
+  scenarioId: string;
+  executableId?: string;
+  model?: FoundationModel;
+}): Promise<AiDeployment> {
+  const query = {
+    scenarioId: opts.scenarioId,
+    status: 'RUNNING' as AiDeploymentStatus,
+    ...(opts.executableId && { executableIds: [opts.executableId] })
+  };
+
+  // TODO: add a cache: https://github.tools.sap/AI/gen-ai-hub-sdk-js-backlog/issues/78
+  let deploymentList: AiDeployment[];
+  const { deploymentQuery } = DeploymentApi;
+  const resourceGroup = { 'AI-Resource-Group': 'default' };
+  try {
+    deploymentList = (await deploymentQuery(query, resourceGroup).execute())
+      .resources;
+  } catch (error) {
+    throw new Error('Failed to fetch the list of deployments: ' + error);
+  }
+
+  if (opts.model) {
+    deploymentList = deploymentList.filter(
+      deployment => extractModel(deployment)?.name === opts.model!.name
```

Comment:
Yes, `tsc` is stupid like that sometimes. This case seems particularly straight forward, therefore I was about to create an issue in TS, but I found that this is another design limitation: `tsc` does not know what else you are doing to that object in your callback (e.g. you could set it to `undefined`), therefore it resets the narrowing you have done before. (https://github.com/microsoft/TypeScript/issues/9998, https://github.com/microsoft/TypeScript/issues/11498)

You can mitigate this by creating a local variable `const model = opts.model;` and work with that. I think it is not worth it.

Generally I dislike the use of `!` as this can be dangerous. Maybe we change code in another place and assert this cannot be `undefined` while it is. In most cases it is better to resort to `?`, because this will resolve to `undefined` and not result in an error at runtime.

```suggestion
      deployment => extractModel(deployment)?.name === opts.model?.name
```

Reply:
I read Matthias comment -> fixed it (after finding out what you already found out 🤦) -> now read your comment 😬, I went with the const model = opts.model approach